# Segmenting and Clustering Neighborhoods in Toronto

### Imports

In [1]:
!pip install bs4
#!pip install requests

  Created wheel for bs4: filename=bs4-0.0.1-py3-none-any.whl size=1272 sha256=f049aef11addd3ac97dca19bdb9650bc5af0fe7889c8e6f56e56b7d04276ce8d
  Stored in directory: /tmp/wsuser/.cache/pip/wheels/0a/9e/ba/20e5bbc1afef3a491f0b3bb74d508f99403aabe76eda2167ca
Successfully built bs4


In [2]:
from bs4 import BeautifulSoup 
import requests  # helps to download a web page

In [3]:
import pandas as pd
import numpy as np

### Getting Data from Wikipedia Page

In [4]:
# retreiving the URL and creating a Beautiful soup object
#The below url contains html tables with data about Toronto neighborhoods
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
data  = requests.get(url).text
soup = BeautifulSoup(data,"html5lib")

In [5]:
#find all html tables in the web page
tables = soup.find_all('table')
len(tables)

3

### Putting Data into a Dataframe

In [6]:
# create a list
table_contents = []

In [7]:
# append table data into list
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned': # Ignore cells with a borough that is Not assigned
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        # If a cell has a borough but a Not assigned  neighborhood, then the neighborhood will be the same as the borough
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

In [8]:
# create a dataframe from list
df = pd.DataFrame(table_contents)

In [9]:
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [10]:
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [11]:
df.shape

(103, 3)

### Getting Latitude and Longitude Data

In [13]:
url2 = 'https://cocl.us/Geospatial_data'
coordinates = pd.read_csv(url2)
print(coordinates)

    Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]


In [14]:
df.columns

Index(['PostalCode', 'Borough', 'Neighborhood'], dtype='object')

In [15]:
lat_long_coords = pd.DataFrame(coordinates)

In [16]:
df.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

In [17]:
lat_long_coords.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

Merge the two tables by matching the Postal Codes

In [21]:
result = pd.merge(df,lat_long_coords,left_on='PostalCode',right_on='Postal Code',how='inner')
result

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood Postal Code   Latitude  \
0                                            Parkwoods         M3A  43.753259   
1                                     Victoria Village         M4A  43.725882   
2                            Regent Park, Harbourfront         M5A  43.654260   
3                     Lawrence Manor, Lawrence Heights         M6A  43.718518   
4                        Ontario Provincial Government         M7A  43.662301   
..                                                 ...         ...        ...   
98       The Kingsway, Montgomery Road, Old Mill North         M8X  43.653654   
99                                Church and Wellesley         M4Y  43.665860   
100                                     Enclave of M4L         M7Y  43.662744   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...         M8Y  43.636258   
102  Mimico NW, The Queensway West, South of Bloor,...         M8Z  43.628841   

     Longitude  
0   -79.329656  
1   -79.315572  
2   -79.360636  
3   -79.464763  
4   -79.389494  
..         ...  
98  -79.506944  
99  -79.383160  
100 -79.321558  
101 -79.498509  
102 -79.520999  

[103 rows x 6 columns]

Drop the duplicate Postal Code column

In [22]:
neighborhoods = result.drop(['Postal Code'],axis=1)
neighborhoods

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [23]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        len(neighborhoods.Neighborhood.unique())
        #neighborhoods.shape[0].unique()
    )
)

The dataframe has 15 boroughs and 103 neighborhoods.


### Create a Map of Toronto

In [25]:
!pip install folium
import folium

     |████████████████████████████████| 94 kB 6.1 MB/s  eta 0:00:01


First, create a map of Toronto visualizing the different neighborhoods and their relative locations

In [26]:
# create map of Toronto using latitude and longitude values
latitude=43.6511
longitude=-79.38
map_toronto = folium.Map(location=[latitude,longitude],zoom_start=10)

# add markers to map
for lat, lng, label in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Import Venue Data

In [27]:
nearby_venues1 = pd.read_json("https://raw.githubusercontent.com/ibm-developer-skills-network/yczvh-DataFilesForIBMProjects/master/segmenting_neighborhoods.json")    
nearby_venues1.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                 'Venue', 
                 'Venue Latitude', 
                 'Venue Longitude', 
                 'Venue Category']
Toronto_venues=nearby_venues1

In [43]:
Toronto_venues.columns

Index(['Neighborhood', 'Neighborhood Latitude', 'Neighborhood Longitude',
       'Venue', 'Venue Latitude', 'Venue Longitude', 'Venue Category'],
      dtype='object')

In [30]:
print('{} venues were returned.'.format(Toronto_venues.shape[0]))

1337 venues were returned.


### Exploring the Neighborhoods in Toronto

In [50]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                               8   
Bathurst Manor, Wilson Heights, Downsview North                     23   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   22   
...                                                                ...   
Willowdale West                                                      5   
Willowdale, Newtonbrook                                              1   
Woburn                                                               4   
Woodbine Heights                                                     8   
York Mills West                                                      2   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                                8   
Bathurst Manor, Wilson Heights, Downsview North                      23   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale West                                                       5   
Willowdale, Newtonbrook                                               1   
Woburn                                                                4   
Woodbine Heights                                                      8   
York Mills West                                                       2   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                               8               8   
Bathurst Manor, Wilson Heights, Downsview North     23              23   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale West                                      5               5   
Willowdale, Newtonbrook                              1               1   
Woburn                                               4               4   
Woodbine Heights                                     8               8   
York Mills West                                      2               2   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                         8   
Bathurst Manor, Wilson Heights, Downsview North               23   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale West                                                5   
Willowdale, Newtonbrook                                        1   
Woburn                                                         4   
Woodbine Heights                                               8   
York Mills West                                                2   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                        8 

Groupby returned 100 rows while total neighborhoods in previous dataframe was 103. Will need to drop NA values when merging towards the end.

In [51]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 241 uniques categories.


### Analyze Each Neighborhood

In [63]:
# one hot encoding
toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['NeighborhoodName'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

NeighborhoodName  Accessories Store  Adult Boutique  \
0                          Malvern, Rouge                  0               0   
1  Rouge Hill, Port Union, Highland Creek                  0               0   
2       Guildwood, Morningside, West Hill                  0               0   
3       Guildwood, Morningside, West Hill                  0               0   
4       Guildwood, Morningside, West Hill                  0               0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Train Station  Truck Stop  \
0                 0                    0  ...              0           0   
1                 0                    0  ...              0           0   
2                 0                    0  ...              0           0   
3                 0                    0  ...              0           0   
4                 0                    0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 242 columns]

In [64]:
# Group rows by neighborhood and take the mean of category frequency
toronto_grouped = toronto_onehot.groupby('NeighborhoodName').mean().reset_index()
toronto_grouped.head()

NeighborhoodName  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  Train Station  \
0              0.0               0.0             0.000000  ...            0.0   
1              0.0               0.0             0.000000  ...            0.0   
2              0.0               0.0             0.000000  ...            0.0   
3              0.0               0.0             0.000000  ...            0.0   
4              0.0               0.0             0.045455  ...            0.0   

   Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  \
0         0.0                            0.0               0.0   
1         0.0                            0.0               0.0   
2         0.0                            0.0               0.0   
3         0.0                            0.0               0.0   
4         0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 242 columns]

In [65]:
toronto_grouped.shape

(100, 242)

### Find the Top 10 most common venues for each neighborhood

In [67]:
# write a function to sort the venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [68]:
# create a new dataframe and display the top 10 venues for each neighborhood
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['NeighborhoodName']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village                       Bank   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot          Skating Rink                Lounge   
1                   Gym          Skating Rink           Coffee Shop   
2                  Bank              Pharmacy    Chinese Restaurant   
3                  Café   Japanese Restaurant    Chinese Restaurant   
4           Coffee Shop    Italian Restaurant              Pharmacy   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Yoga Studio      Department Store     Electronics Store   
1                   Pub        Sandwich Place              Pharmacy   
2         Shopping Mall           Bridal Shop        Sandwich Place   
3           Yoga Studio    Dim Sum Restaurant                 Diner   
4       Thai Restaurant             Juice Bar               Butcher   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant             Drugstore             Donut Shop  
1                  Gas Station   Distribution Center              Gastropub  
2                Deli / Bodega            Restaurant            Supermarket  
3               Discount Store   Distribution Center       Department Store  
4            Indian Restaurant      Sushi Restaurant             Restaurant

### Cluster Neighborhoods

In [70]:
from sklearn.cluster import KMeans

In [71]:
# Run k-means to cluster the neighborhoods into 5 clusters
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('NeighborhoodName', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 3, 3, 0, 3, 0, 0, 0, 0, 0], dtype=int32)

In [83]:
kmeans.labels_.shape

(100,)

In [72]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhoods

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656             1.0                  Park       Food & Drink Shop   
1 -79.315572             3.0     French Restaurant             Pizza Place   
2 -79.360636             0.0           Coffee Shop                    Park   
3 -79.464763             3.0        Clothing Store  Furniture / Home Store   
4 -79.389494             3.0           Coffee Shop                   Diner   

  3rd Most Common Venue  4th Most Common Venue  5th Most Common Venue  \
0           Yoga Studio          Deli / Bodega      Electronics Store   
1           Coffee Shop  Portuguese Restaurant           Intersection   
2        Breakfast Spot                   Café                 Bakery   
3     Accessories Store            Coffee Shop  Vietnamese Restaurant   
4      Sushi Restaurant    Distribution Center         Sandwich Place   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Eastern European Restaurant             Drugstore            Donut Shop   
1                 Hockey Arena    Dim Sum Restaurant          Dance Studio   
2               Chocolate Shop   Distribution Center                   Spa   
3                     Boutique   Distribution Center          Dessert Shop   
4                   Restaurant                  Park    Mexican Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0               Dog Run    Distribution Center  
1         Deli / Bodega       Department Store  
2          Dessert Shop         Farmers Market  
3    Dim Sum Restaurant                  Diner  
4   Japanese Restaurant     Italian Restaurant

In [85]:
toronto_merged.tail()

PostalCode                Borough  \
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944   
99                                Church and Wellesley  43.665860 -79.383160   
100                                     Enclave of M4L  43.662744 -79.321558   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509   
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999   

     Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
98              0.0                  Pool            Smoke Shop   
99              0.0           Coffee Shop                  Café   
100             0.0           Pizza Place            Smoke Shop   
101             4.0        Baseball Field           Yoga Studio   
102             0.0                   Gym           Flower Shop   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
98                  River           Yoga Studio          Dance Studio   
99    Japanese Restaurant      Sushi Restaurant      Ramen Restaurant   
100            Skate Park    Light Rail Station                  Park   
101    Falafel Restaurant           Escape Room     Electronics Store   
102           Social Club        Discount Store            Kids Store   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
98   Eastern European Restaurant             Drugstore            Donut Shop   
99           Martial Arts School     Indian Restaurant            Steakhouse   
100                      Brewery        Farmers Market  Fast Food Restaurant   
101  Eastern European Restaurant             Drugstore            Donut Shop   
102         Fast Food Restaurant        Sandwich Place          Burger Joint   

    9th Most Common Venue 10th Most Common Venue  
98                Dog Run    Distribution Center  
99   Ethiopian Restaurant     Italian Restaurant  
100         Burrito Place                Butcher  
101               Dog Run    Distribution Center  
102         Burrito Place                 Bakery

Drop the NA values in the cluster labels that converted column 'Cluster Labels' into float data type

In [88]:
toronto_merged = toronto_merged.dropna()

In [89]:
toronto_merged = toronto_merged.astype({'Cluster Labels': int})

Now to visualize the resulting clusters:

In [76]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [91]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters